File: fast_mm_matlab/src/utils/JointToMargin.m
% This function inputs some information of joint distribution
% support space, and output three (n_col) * (sum_col + 1) matrices,
% which are the coefficients to transform the joint distribution
% to the marginal distributions.

function joint_to_margin = JointToMargin(column, sum_col)
  n_col = size(column, 1);
  joint_to_margin = cell(1, 3);
  joint_to_margin{1} = sparse(n_col, sum_col + 1);
  joint_to_margin{2} = sparse(n_col, sum_col + 1);
  joint_to_margin{3} = sparse(n_col, sum_col + 1);
  for i = 1 : n_col
    joint_to_margin{1}(i, column(i, 1) + 1) = 1;
    joint_to_margin{2}(i, column(i, 2) + 1) = 1;
    joint_to_margin{3}(i, column(i, 3) + 1) = 1;
  end
end

File: fast_mm_python/src/utils/JointToMargin.py
"""
Generate transformation matrices to convert joint distribution to marginal distributions.
"""

import numpy as np
from scipy import sparse


def JointToMargin(column, sum_col):
    """
    Generate transformation matrices to convert joint distribution to marginal distributions.
    
    Args:
        column: Array of shape (n_col, 3) representing distribution support space
        sum_col: Sum value for the columns
    
    Returns:
        joint_to_margin: List of 3 sparse matrices, each of shape (n_col, sum_col+1)
    """
    n_col = column.shape[0]
    joint_to_margin = []
    
    for dim in range(3):
        matrix = sparse.lil_matrix((n_col, sum_col + 1), dtype=np.float64)
        for i in range(n_col):
            matrix[i, column[i, dim]] = 1.0
        joint_to_margin.append(matrix.tocsr())
    
    return joint_to_margin

